cpp/src/arrow/engine/substrait/options.cc

```diff
@@ -0,0 +1,99 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+#include <iostream>
+
+#include "arrow/engine/substrait/options.h"
+
+#include <google/protobuf/util/json_util.h>
+#include "arrow/compute/exec/asof_join_node.h"
+#include "arrow/compute/exec/options.h"
+#include "arrow/engine/substrait/expression_internal.h"
+#include "arrow/engine/substrait/relation_internal.h"
+#include "substrait/extension_rels.pb.h"
+
+namespace arrow {
+namespace engine {
+
+class DefaultExtensionProvider : public ExtensionProvider {
+ public:
+  Result<DeclarationInfo> MakeRel(const std::vector<DeclarationInfo>& inputs,
+                                  const google::protobuf::Any& rel,
+                                  const ExtensionSet& ext_set) override {
+    if (rel.Is<arrow::substrait::AsOfJoinRel>()) {
+      arrow::substrait::AsOfJoinRel as_of_join_rel;
+      rel.UnpackTo(&as_of_join_rel);
+      return MakeAsOfJoinRel(inputs, as_of_join_rel, ext_set);
+    }
+    return Status::NotImplemented("Unrecognized extension in Susbstrait plan: ",
+                                  rel.DebugString());
+  }
+
+ private:
+  Result<DeclarationInfo> MakeAsOfJoinRel(
+      const std::vector<DeclarationInfo>& inputs,
+      const arrow::substrait::AsOfJoinRel& as_of_join_rel, const ExtensionSet& ext_set) {
+    if (inputs.size() < 2) {
+      return Status::Invalid("substrait::AsOfJoinNode too few input tables: ",
+                             inputs.size());
+    }
+    // on-key
+    if (!as_of_join_rel.has_on()) {
+      return Status::Invalid("substrait::AsOfJoinNode missing on-key");
+    }
+    ARROW_ASSIGN_OR_RAISE(auto on_key_expr, FromProto(as_of_join_rel.on(), ext_set, {}));
+    if (on_key_expr.field_ref() == NULLPTR) {
+      return Status::NotImplemented("substrait::AsOfJoinNode non-field-ref on-key");
+    }
+    const FieldRef& on_key = *on_key_expr.field_ref();
+
+    // by-key
+    std::vector<FieldRef> by_key;
```

Comment:
I think `by_keys` here is a little clear since this is a vector

Reply:
As discussed, the vector refers to columns, so it is a single "by" key.